test/runtime/netlink.c

```diff
@@ -287,16 +287,19 @@ static void test_getaddr(void)
     msg.msg_controllen = 0;
     msg.msg_flags = 0;
 
-    req.msg.rtgen_family = AF_INET;
-    memcpy(iov.iov_base, &req, sizeof(req));
-    iov.iov_len = sizeof(req);
-    ret = sendmsg(fd, &msg, 0);
-    test_assert(ret == sizeof(req));
-    iov.iov_len = sizeof(buf);
-    recv_resp(fd, &msg, sizeof(struct ifaddrmsg), RTM_NEWADDR, NLM_F_MULTI);
-    ifa = (struct ifaddrmsg *)NLMSG_DATA(buf);
-    test_assert(ifa->ifa_family == AF_INET);
-
+    const int afs[3] = { AF_INET, AF_INET6, AF_UNSPEC };
+    for (int i = 0; i < 3; i++) {
+        req.msg.rtgen_family = afs[i];
+        memcpy(iov.iov_base, &req, sizeof(req));
+        iov.iov_len = sizeof(req);
+        ret = sendmsg(fd, &msg, 0);
+        test_assert(ret == sizeof(req));
+        iov.iov_len = sizeof(buf);
+        recv_resp(fd, &msg, sizeof(struct ifaddrmsg), RTM_NEWADDR, NLM_F_MULTI);
```

Comment:
This makes the test fail when run on Linux, with "Error: ret >= NLMSG_LENGTH(len) -- failed at /usr/src/nanos/test/runtime/netlink.c:68".
Specifically, for multipart messages such as the response to this RTM_GETADDR request there is no guarantee that all chunks are received in a single recvmsg() call: in my Linux box, the NLMSG_DONE chunk is not included in the data received by the first recvmsg() call, and is returned in the following call, which makes the test fail at the second iteration of the for() loop.
One way to get around this issue would be to use different netlink sockets to make different requests.

Reply:
Ok. I changed the loop to open and close a netlink socket for each request. The test now passes under Linux.